File: BOJ/Gold V/BOJ1753.py
#BOJ1753 최단경로 20210312
import sys
import heapq
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

INF = 2**30

def dijkstra(adj, start, weights):
    heap = []
    heapq.heappush(heap, [weights[start], start])
    while heap:
        weight, v = heapq.heappop(heap)
        print(weight,v)
        if weights[v] < weight: continue
        for nv, w in adj[v]:
            if weights[nv] > weight + w:
                weights[nv] = weight + w
                heapq.heappush(heap, [weights[nv], nv])

def main():
    n,m = map(int, input().rstrip().split())
    start = int(input())
    adj = [ [] for _ in range(n+1)]
    weights = [INF for _ in range(n+1)]
    weights[start] = 0
    for _ in range(m):
        u,v,w = map(int, input().rstrip().split())
        adj[u].append((v,w))
    dijkstra(adj, start, weights)
    for i in range(1,n+1):
        v = weights[i]
        print(v if v != INF else "INF")
if __name__ == '__main__':
    main()

File: BOJ/Gold V/BOJ1753.java
//BOJ1753 최단경로 20210713
import java.io.*;
import java.util.*;

import static java.lang.Math.*;

public class Main {
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private static int INF = Integer.MAX_VALUE;

    private static int n,m;
    private static List<List<int[]>> adj;
    private static int[] ans;


    public static void main(String[] args) throws IOException {
        StringTokenizer stk = new StringTokenizer(br.readLine());
        n = nextInt(stk);
        m = nextInt(stk);
        
        int start = Integer.parseInt(br.readLine());
        ans = new int[n+1];
        for(int i = 1; i <= n; i++){
            ans[i] = INF;
        }

        adj = new ArrayList<>();
        for(int i = 0; i <= n; i++){
            adj.add(new ArrayList<>());
        }

        for(int i = 0; i < m; i++){
            stk = new StringTokenizer(br.readLine());
            int u = nextInt(stk), v = nextInt(stk), w = nextInt(stk);
            adj.get(u).add(new int[]{v, w});
        }

        dijkstra(start);
        for(int i = 1; i <= n; i++){
            bw.write(((ans[i] == INF) ? "INF" : ans[i]) + "\n");
        }

        bw.flush();
        bw.close();
    }

    public static void dijkstra(int start){
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(x -> x[1]));
        ans[start] = 0;
        pq.add(new int[]{start, 0});

        while(!pq.isEmpty()){
            int[] p = pq.poll();
            int v = p[0], w = p[1];
            if(w > ans[v]) continue;

            for(int[] nv: adj.get(v)){
                if(nv[1] + w < ans[nv[0]]){
                    ans[nv[0]] = nv[1] + w;
                    pq.add(new int[]{nv[0], (nv[1]+w)});
                }
            }
        }
    }

    public static int nextInt(StringTokenizer stk){
        return Integer.parseInt(stk.nextToken());
    }
}